### PR TITLE
Support for Nested FilteredRelation

### PR DESCRIPTION
**Pull request by ferrants**
https://code.djangoproject.com/ticket/29789

There are some other places I considered making changes, noted in the comments. I'd remove these comments before wanting to merge, but left them there for conversation sake.

I also have a branch that does this for 3.0, some small differences, mostly in the `resolve_expression` function signature. I can open the PR for 3.0 as well, but I need the 2.2 one for work I am actively doing.

### CODE REVIEW

File: django/db/models/query_utils.py [hunk 1]
```diff
@@ -332,5 +335,9 @@ def resolve_expression(self, *args, **kwargs):
     def as_sql(self, compiler, connection):
         # Resolve the condition in Join.filtered_relation.
         query = compiler.query
-        where = query.build_filtered_relation_q(self.condition, reuse=set(self.path))
+        # Add other usable aliases in the query to the reuse set.
+        # Check for if it can be used is in Query.join
+        reusable_aliases = self.path
+        reusable_aliases += list(query.alias_map.keys())
```
Comment (ferrants) [inline django/db/models/query_utils.py]: This change allows lookups to use relations not in the direct mapping of the FilteredRelation

File: django/db/models/sql/query.py [hunk 2]
```diff
@@ -1388,7 +1382,15 @@ def names_to_path(self, names, opts, allow_many=True, fail_on_missing=False):
...
                 elif name in self._filtered_relations and pos == 0:
                     filtered_relation = self._filtered_relations[name]
-                    field = opts.get_field(filtered_relation.relation_name)
+                    if LOOKUP_SEP in filtered_relation.relation_name:
+                        parts = filtered_relation.relation_name.split(LOOKUP_SEP)
+                        filtered_relation_path, field, _, _ = self.names_to_path(
+                            parts, opts, allow_many, fail_on_missing
+                        )
+                        # don't want the last one, because that one gets added on below and would be duplicate
+                        path.extend(filtered_relation_path[:-1])
```
Comment (ferrants) [inline django/db/models/sql/query.py]: This allows the nested FilterRelation `relation_name`

**Comment by felixxm** [comment]
Closing per [comment](https://code.djangoproject.com/ticket/29789#comment:11), please create PR targeted to the master branch.

**Comment by ferrants** [comment]
Opened https://github.com/django/django/pull/12293 against master

**Comment by sshivananda** [comment]
@ferrants  do you know what versions of django is this feature available in ?
(hoping to use this in 2.2 )

**Comment by felixxm** [comment]
@sshivananda This is available in Django 3.2, see 7d6916e82700896ef23c50720d8cf0de4d6b9074.

**Pull request closed without merging**